kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,472 @@
+ */
+
+package dev.responsive.kafka.internal.db.mongo;
+
+import static com.mongodb.MongoClientSettings.getDefaultCodecRegistry;
+import static dev.responsive.kafka.internal.db.partitioning.SegmentPartitioner.UNINITIALIZED_STREAM_TIME;
+import static dev.responsive.kafka.internal.stores.ResponsiveStoreRegistration.NO_COMMITTED_OFFSET;
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
+import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.FindOneAndUpdateOptions;
+import com.mongodb.client.model.ReturnDocument;
+import com.mongodb.client.model.UpdateOneModel;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.model.Updates;
+import com.mongodb.client.model.WriteModel;
+import dev.responsive.kafka.internal.db.MongoWindowFlushManager;
+import dev.responsive.kafka.internal.db.RemoteWindowedTable;
+import dev.responsive.kafka.internal.db.partitioning.SegmentPartitioner;
+import dev.responsive.kafka.internal.db.partitioning.SegmentPartitioner.SegmentPartition;
+import dev.responsive.kafka.internal.db.partitioning.SegmentPartitioner.SegmentRoll;
+import dev.responsive.kafka.internal.stores.RemoteWriteResult;
+import dev.responsive.kafka.internal.utils.Iterators;
+import dev.responsive.kafka.internal.utils.SegmentBatch;
+import dev.responsive.kafka.internal.utils.WindowedKey;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.pojo.PojoCodecProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MongoWindowedTable implements RemoteWindowedTable<WriteModel<WindowDoc>> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MongoWindowedTable.class);
+  private static final String METADATA_COLLECTION_NAME = "window_metadata";
+
+  private static final UpdateOptions UPSERT_OPTIONS = new UpdateOptions().upsert(true);
+
+  private final String name;
+  private final SegmentPartitioner partitioner;
+
+  // whether to put windowStartMs first in the composite windowed key format in WindowDoc
+  private final boolean timestampFirstOrder;
+
+  private final MongoDatabase database;
+  private final MongoCollection<WindowMetadataDoc> metadata;
+  private final ConcurrentMap<Integer, PartitionSegments> kafkaPartitionToSegments =
+      new ConcurrentHashMap<>();
+
+  private class PartitionSegments {
```

Comment:
I think it would be easier to read if this were a static class and you pass in the reference to `database`

Reply:
Will do (also open to naming suggestions for any of these classes, I think better names would help but I can only come up with so many names lol)